bin/src/lib.rs

```diff
@@ -840,3 +871,95 @@ mod tests {
         }
     }
 }
+
+#[cfg(feature = "metrics")]
+pub struct ConfigMetrics {
+    zones_file: Counter,
+    #[cfg(feature = "sqlite")]
+    zones_sqlite: Counter,
+    #[cfg(feature = "resolver")]
+    zones_forwarder: Counter,
+
+    zones_primary_file: Counter,
+    zones_secondary_file: Counter,
+    #[cfg(feature = "sqlite")]
+    zones_primary_sqlite: Counter,
+    #[cfg(feature = "sqlite")]
+    zones_secondary_sqlite: Counter,
+}
+
+#[cfg(feature = "metrics")]
+impl ConfigMetrics {
+    pub fn new(config: &Config) -> Self {
+        let hickory_info = counter!("hickory_info", "version" => hickory_client::version());
+        describe_counter!("hickory_info", Unit::Count, "hickory service metadata");
+        hickory_info.increment(1);
+
+        let hickory_config_info = counter!("hickory_config_info",
+            "directory" => config.directory().to_string_lossy().to_string(),
+            "disable_https" => config.disable_https().to_string(),
+            "disable_quic" => config.disable_quic().to_string(),
+            "disable_tcp" => config.disable_tcp().to_string(),
+            "disable_tls" => config.disable_tls().to_string(),
+            "disable_udp" => config.disable_udp().to_string(),
+            "allow_networks" => config.allow_networks().len().to_string(),
+            "deny_networks" => config.deny_networks().len().to_string(),
+            "zones" => config.zones().len().to_string()
+        );
+        describe_counter!(
+            "hickory_config_info",
+            Unit::Count,
+            "hickory config metadata"
+        );
+        hickory_config_info.increment(1);
+
+        let zones_total_name = "hickory_zones_total";
+        let zones_primary_name = "hickory_zones_primary_total";
+        let zones_secondary_name = "hickory_zones_secondary_total";
+
+        let zones_file = counter!(zones_total_name, "store" => "file");
+        let zones_primary_file = counter!(zones_primary_name, "store" => "file");
+        let zones_secondary_file = counter!(zones_secondary_name, "store" => "file");
```

Comment:
We could simplify this by collapsing the primary/secondary metrics into `hickory_zones_total`, and representing the primary/secondary role as another attribute on it. (same for the SQLite store)
```suggestion
        let zones_total_name = "hickory_zones_total";

        let zones_primary_file = counter!(zones_total_name, "store" => "file", "role" => "primary");
        let zones_secondary_file = counter!(zones_total_name, "store" => "file", "role" => "secondary");
```

Reply:
That's fine. Will change the primary/secondary to an attribute.